packages/opentelemetry-core/src/trace/instrumentation/BasePlugin.ts

```diff
@@ -15,13 +15,34 @@
  */
 
 import { Tracer, Plugin, Logger } from '@opentelemetry/types';
+import * as semver from 'semver';
+import * as path from 'path';
+
+/**
+ * Maps a name (key) representing a internal file module and its exports
+ */
+export interface ModuleNameToFilePath {
+  [moduleName: string]: string; // moduleName: path/to/file
+}
+
+/**
+ * Maps a semver to a module:filepath Map
+ */
+export interface ModuleExportsMapping {
+  [semver: string]: ModuleNameToFilePath;
+}
 
 /** This class represent the base to patch plugin. */
 export abstract class BasePlugin<T> implements Plugin<T> {
+  supportedVersions?: string[];
   protected _moduleExports!: T;
   protected _tracer!: Tracer;
   protected _logger!: Logger;
-  supportedVersions?: string[];
+  protected readonly _internalFilesList?: ModuleExportsMapping
+  protected _internalFilesExports!: { [module: string]: any };
+  protected basedir!: string;
+
+  constructor(readonly moduleName?: string, readonly version?: string) {}
```

Comment:
This is how constructors were implemented in grpc/http plugin. Plugins would need to pass in versions/moduleNames/basedir in constructor and `basedir` is passed in via config to `enable`. I think its confusing passing  in `constructor`, some parts in `enable`, so I want to start a discussion on how to make this more clear

I decided not to include private member `setPluginContext` since it would only be used by this stuff currently.
https://github.com/census-instrumentation/opencensus-node/blob/ef5712fd3b279b0e80494322231232047b06f9e6/packages/opencensus-core/src/trace/instrumentation/base-plugin.ts#L74-L82